Dashboard refactory

graceguo-supercat opened the pull request:
done:

- Create DashboardViewContainer component, the root component for Dashboard.
- Create ChartContainer component (with Chart and ChartBody), it has chartActions and chartReducers.
- Charts reducer is global store for all charts data. it is a map and each entry represents the status/data of a chart.
- we access single chart data, dispatch actions, or update chart status, with chartKey (string format slice_id).
- dashboard view, same as explore view, we use dashboard reducer hold dashboard data (like grid layout, dashboard title, and slice title etc), and charts reducer only hold charts data.
- Use common components in the whole code base, like FavStart, TooltipWrapper, etc.
- use redux action for add, remove, and apply filter
- timeout and other errors handling, click to display stack trace (same like explore view)

Some changes according to recent code review:
- linting fix: rename viz_type to vizType, use spread operator replace Object.assign, etc;
- no binding in callback props;
- resize dashboard should only re-render charts without refetch;
- use Loading component when chart is loading;
- for each slice in dashboard, when user clicks edit chart or explore chart, open a new window instead of use current window.

Commit:
add redux store and init state

Commit:
dashboard container and other components

michellethomas commented on superset/assets/javascripts/dashboard/components/SliceInner.jsx:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  show() {
+    //
```
Should this be empty?

mistercrunch commented on superset/assets/javascripts/dashboard/actions.js:
```diff
@@ -0,0 +1,171 @@
+        // if server-side reject the overwrite action,
+        // revert to old state
+        dispatch(updateSliceName(slice, oldName));
+        notify.error('You don\'t have the rights to alter this slice');
```
NIT: when strings contain `'`, it's nice to quote them with `"`

graceguo-supercat replied:
fixed

mistercrunch commented on superset/assets/javascripts/dashboard/actions.js:
```diff
@@ -0,0 +1,171 @@
+  return { type: FETCH_SLICE_TIMEOUT, slice, timeout };
+}
+
+export function fetchSlice(slice, sliceUrl, timeout) {
```
We should share code with this location here:
https://github.com/apache/incubator-superset/blob/master/superset/assets/javascripts/explore/actions/chartActions.js#L51

Arguably all of chartActions and chartReducer should be in sync here.

Ideally the components, actions, reducers and state are all bound together and reused in all places. The idea of refactoring `chartActions` out of `explore/actions` and `chartReducer` was that it could be reused here and wherever the ChartComponent is used

graceguo-supercat replied:
please see updated chart component.

mistercrunch commented on superset/assets/javascripts/dashboard/components/SliceContainer.jsx:
```diff
@@ -0,0 +1,198 @@
+  refresh: PropTypes.bool,
+};
+
+class SliceContainer extends React.PureComponent {
```
About naming and modeling of components here, I think ideally we'd have something like 
* `SliceContainer`, that is simple as possible and only does the redux interaction and passes on props to
* `Slice` which represents the whole slice, header and body
* `SliceBody` which is now called `Slice`

graceguo-supercat replied:
Please see the latest commit. Now i wrap name component like `ChartContainer`, `Chart`, and `ChartBody`. 
`ChartContainer` is redux container.
`Chart` only responsible for chart rendering part. It's parent component, either dashboard or explore view, provide header controls.
`ChartBody` provides dom operation wrapper called from d3.

mistercrunch commented on superset/assets/javascripts/dashboard/components/SliceHeader.jsx:
```diff
@@ -0,0 +1,120 @@
+  forceRefresh: PropTypes.func,
+};
+
+const SliceHeader = ({ removeSlice, updateSliceName, toggleExpandSlice, forceRefresh,
```
This seems a bit big for a function component. My assumption is that those always fully render on any parent render, as opposed to `PureComponent` which only render only if state or props have shallow-changed.  I'm not sure if this is a valid concern though.

mistercrunch commented on superset/assets/javascripts/dashboard/components/DashboardViewContainer.jsx:
```diff
@@ -0,0 +1,335 @@
+  isStarred: false,
+};
+
+class DashboardViewContainer extends React.PureComponent {
```
I think here too is would be good to de-couple the `DashboardContainer` (manages redux, and passes props along to `Dashboard`).

graceguo-supercat replied:
fixed. now DashboardContainer.jsx is redux only component.

Commit:
1. fix pre-loaded filter populate data
2. fix inline edit css

Commit:
share chart component between dashboard and explore view

1. create ChartContainer component, with chartActions and chartReducers
2. charts reducer contains a map of chart, each entry represents the status/data of a chart
3. in both dashboard and explore view, access chart data by chartKey(slice_id string)
4. dashboardview container, same as exploreview container, we use dashboard reducer hold dashboard data (like grid layout, dashboard title, and slice title etc), and charts reducer only hold charts data.
5. separated dashboard presentation componenent, and dashboardview container for redux container.

graceguo-supercat commented on superset/assets/javascripts/dashboard/components/Slice.jsx:
```diff
@@ -0,0 +1,113 @@
+    super(props);
+
+    // these properties are used by visualizations
+    this.containerId = props.containerId;
```
because d3 called our component with `instance method` way. while react component doesn't expose methods like that...

Commit:
Merge branch 'master' into gg-ChartComponent

mistercrunch commented on superset/assets/javascripts/chart/ChartContainer.jsx:
```diff
@@ -0,0 +1,151 @@
+    const loading = this.props.chartStatus === 'loading';
+    const containerId = this.props.containerId;
+    return (
+      <div>
```
Personally I'd rather have the `loading`-related logic in the `Chart` component than in `ChartContainer`. As much as possible `ChartContainer` should only manage the Redux interactions.

That applies to `<StackTraceMessage/>` as well as whatever else isn't Redux-related

mistercrunch commented on superset/assets/javascripts/chart/Chart.jsx:
```diff
@@ -0,0 +1,115 @@
+/* eslint camelcase: 0 */
```
I like the folder structure here with everything chart-related under a `chart/` folder

mistercrunch commented on superset/assets/javascripts/components/StackTraceMessage.jsx:
```diff
@@ -0,0 +1,73 @@
+    };
+  }
+
+  removeAlert() {
```
I think the feature allowing to remove alerts is kind of legacy (not sure how it got to be a requirement in the first place). I think ExploreV2 was trying to replicate an MVC legacy feature. Anyhow we could remove that (and related actions/reducers) as part of this PR or later if you prefer. I'm just putting this on the radar here.

graceguo-supercat replied:
fixed. removed removeAlerts feature.

mistercrunch commented on superset/assets/javascripts/dashboard/components/DashboardViewContainer.jsx:
```diff
@@ -0,0 +1,305 @@
+import urlLib from 'url';
+
+import * as dashboardActions from '../actions';
+import * as chartActions from '../../chart/chartAction';
```
happy to see that done properly here!

mistercrunch commented on superset/assets/javascripts/dashboard/components/DashboardViewContainer.jsx:
```diff
@@ -0,0 +1,305 @@
+    return JSON.stringify(this.props.filters, null, '  ');
+  }
+
+  updateDashboardTitle(title) {
```
I'm trying to understand why these methods that match actions 1:1 are needed? Seems like this is boilerplate/unneeded redirection...

mistercrunch commented on superset/assets/javascripts/dashboard/components/DashboardViewContainer.jsx:
```diff
@@ -0,0 +1,305 @@
+        removeChart={this.removeChart.bind(this)}
+        updateDashboardLayout={this.updateDashboardLayout.bind(this)}
+        addSlicesToDashboard={this.addSlicesToDashboard.bind(this)}
+        toggleExpandSlice={this.toggleExpandSlice.bind(this)}
```
couldn't you just:
`toggleExpandSlice={this.props.actions.toggleExpandSlice}` here?

Also to avoid much boilerplate here I wouldn't be against having something like `dashboardActions` and pass that to `<Dashboard>`. Could also go in the direction of reusable and composable actions/reducers

graceguo-supercat replied:
Fixed. I prefer to use `Callback Functions as Props` instead of passing action creators itself down to children components. It will be easier to test child component by just mocking callbacks. Also It's one of good practice mentioned in Airbnb redux/best-practices, as well as outside airbnb:
http://www.thegreatcodeadventure.com/the-react-plus-redux-container-pattern/

mistercrunch commented on superset/assets/javascripts/dashboard/components/GridCell.jsx:
```diff
@@ -0,0 +1,108 @@
+    return (
+      <div className="slice-cell" id={`${slice.slice_id}-cell`}>
+        <div ref={this.getHeaderId(slice)}>
+          <SliceHeader
```
So we have 2 types of chart headers (one in explore and one in dashboard) and eventually I think that should live inside `<Chart/>` as opposed to in those respective apps.

I'm thinking we'd add a `showHeader` bool prop to `Chart` as well as and `showExtendedHeader` or something like that.

We can save that for a future refactor if you want.

graceguo-supercat replied:
It's hard to find common functions or common UI between slice header in dashboard and chart header in explore view :(
I assume we will have a UI/UX redesign very soon. we can provide different header in different `mode` together with redesign.

mistercrunch commented on superset/assets/javascripts/dashboard/components/GridCell.jsx:
```diff
@@ -0,0 +1,108 @@
+            id={'token_' + slice.slice_id}
+            className={`token col-md-12 ${this.props.isLoading ? 'is-loading' : ''}`}
+          >
+            <ChartContainer
```
I'm surprised to see this here. I was thinking that is would go `DashboardContainer -> Dashboard -> Chart`, meaning the entry point for Redux would be at the top only. I'm not quite sure on the implications of this or whether it's the desired model, maybe @williaster can chime in here.

ljharb replied:
Generally you can go with either "only redux at the top", or "redux in every component that needs it". The downsides of the former is that every dependency of the bottom of the chain has to be passed down through every item in the chain - which forces the intervening components to have knowledge of the implementation details of the leaves.

mistercrunch replied:
@ljharb in this context we have a chart component that has a related list of actions and reducers that we reuse across apps, does that inform the Redux injection pattern in any way?

ljharb replied:
Somewhat; if the store is injected via context, then the chart component can use the same global store, which can vary across apps - it definitely means, however, that you can't statically import the store in the leaves.

graceguo-supercat replied:
`dashboardcontainer`'s root reducer is a combined reducer with `dashboard` reducer and `chart` reducer. I want to use `dashboard` reducer only handle dashboard state, like dashboard layout, each slice title/description data, and use `chart` reducer only handle each chart rendering data. The problem bothered me is, chart is deep nested inside Dashboard component. And middle-level components (`GridLayout`, `GridCell`) has nothing more than passing down chart data.

mistercrunch commented on superset/assets/javascripts/dashboard/components/GridLayout.jsx:
```diff
@@ -14,117 +12,94 @@ const ResponsiveReactGridLayout = WidthProvider(Responsive);
+  timeout: PropTypes.number,
+  onChange: PropTypes.func,
+  getFormDataExtra: PropTypes.func,
+  fetchSlice: PropTypes.func,
```
a `dashboardActions` object would remove a bunch of boilerplate here I think

williaster replied:
it's harder to lint actions not being passed with this approach, unless you defined a strict `PropTypes.shape` was defined

mistercrunch commented on superset/assets/javascripts/dashboard/components/SliceAdder.jsx:
```diff
@@ -11,6 +11,8 @@ require('react-bootstrap-table/css/react-bootstrap-table.css');
 const propTypes = {
   dashboard: PropTypes.object.isRequired,
   triggerNode: PropTypes.node.isRequired,
+  user_id: PropTypes.string.isRequired,
```
Surprised the linter doesn't complain here...

graceguo-supercat replied:
now i renamed all user_id to userId.

mistercrunch commented on superset/assets/javascripts/explore/components/ChartContainer.jsx:
```diff
@@ -1,362 +0,0 @@
-import $ from 'jquery';
```
no biggy, but if you had used `git mv` instead of `mv` it would have made the PR a bit more aligned.

mistercrunch commented on superset/assets/javascripts/dashboard/components/DashboardViewContainer.jsx:
```diff
@@ -0,0 +1,305 @@
+  isStarred: PropTypes.bool,
+};
+
+class DashboardViewContainer extends React.PureComponent {
```
Why naming `DashboardViewContainer` instead of just `DashboardContainer` ?

graceguo-supercat replied:
renamed. we now have `DashboardContainer`. but still use `ExploreViewContainer`. We will have redesign for Explore view very soon.

mistercrunch commented on superset/assets/javascripts/explore/actions/chartActions.js:
```diff
@@ -1,74 +0,0 @@
-import { getExploreUrl } from '../exploreUtils';
```
`git mv` > `mv`

mistercrunch commented on superset/assets/javascripts/explore/components/ExploreChartPanel.jsx:
```diff
@@ -0,0 +1,181 @@
+        <Panel
+          style={{ height: this.props.height }}
+          header={
+            <div
```
(optional) we should refactor a `ExploreChartHeader` out as some point.

graceguo-supercat replied:
done.

mistercrunch reviewed (changes requested):
We'll be in such a better place once this lands!

Commit:
fix per code review comments

- remove boiler-plate in dashboard container
- rename dashboardviewcontainer to dashboardcontainer
- remove removeAlert feature from StackTraceMessage
- user userId instead of user_id
- wrapping isLoading related logic inside Chart component
- add ExploreChartHeader component
- we should pre-populate filters from dashboard.metadata

Commit:
Merge remote-tracking branch 'supercat/gg-ChartComponent' into gg-ChartComponent

Commit:
apply airbnb redux good practices for dashboard and chart container component:

- Only containers should call connect. No connect calls in /Components.
- Containers should not create any new Components inside the file.

williaster commented on superset/assets/javascripts/chart/Chart.jsx:
```diff
@@ -0,0 +1,190 @@
+    const isLoading = this.props.chartStatus === 'loading';
+    return (
+      <div className={`token col-md-12 ${isLoading ? 'is-loading' : ''}`}>
+        {isLoading && <img
```
nit -- you're not consistent about new lines with this style of rendering (see line 164 below). I would always but the component on a new line and indent. we should enable this lint rule!

graceguo-supercat replied:
i can add this rule into our .eslintrc

williaster commented on superset/assets/javascripts/chart/Chart.jsx:
```diff
@@ -0,0 +1,190 @@
+        />}
+
+        {this.props.chartAlert &&
+        <StackTraceMessage
```
nit -- indent

williaster commented on superset/assets/javascripts/chart/Chart.jsx:
```diff
@@ -0,0 +1,190 @@
+        }
+
+        {!this.props.chartAlert &&
+        <ChartBody
```
nit -- indent

williaster commented on superset/assets/javascripts/chart/Chart.jsx:
```diff
@@ -0,0 +1,190 @@
+        <ChartBody
+          containerId={this.containerId}
+          vizType={this.props.formData.viz_type}
+          height={this.height.bind(this)}
```
why do you bind this in the render vs all the other methods in the constructor? you shouldn't bind in a render because it defines a new func every render

williaster commented on superset/assets/javascripts/chart/ChartBody.jsx:
```diff
@@ -0,0 +1,54 @@
+const propTypes = {
+  containerId: PropTypes.string,
+  vizType: PropTypes.string,
+  height: PropTypes.func,
```
I'd make these required or add defaults, if they're not passed now your `this.props.width` etc calls below would throw

williaster commented on superset/assets/javascripts/chart/chartAction.js:
```diff
@@ -0,0 +1,90 @@
+export function runQuery(formData, force = false, timeout = 60, key) {
+  return (dispatch) => {
+    const url = getExploreUrl(formData, 'json', force);
+    const queryRequest = $.ajax({
```
is there a reason we don't use fetch over ajax @mistercrunch? not in the scope of this PR just curious.

graceguo-supercat replied:
I guess the early stage of Superset project is all jQuery...

williaster commented on superset/assets/javascripts/chart/chartReducer.js:
```diff
@@ -0,0 +1,108 @@
+export default function chartReducer(charts = {}, action) {
+  const actionHandlers = {
+    [actions.CHART_UPDATE_SUCCEEDED](state) {
+      return Object.assign(
```
why do you not use object rest spread over `Object.assign`? I think generally this is preferred.
eg `{ ...state, { chartStatus, queryResponse, chartUpdateEndTime } }`

williaster commented on superset/assets/javascripts/chart/chartReducer.js:
```diff
@@ -0,0 +1,108 @@
+      );
+    },
+    [actions.CHART_UPDATE_STARTED](state) {
+      return Object.assign({}, state,
```
ditto object spread `{ ... }`

williaster commented on superset/assets/javascripts/chart/chartReducer.js:
```diff
@@ -0,0 +1,108 @@
+        });
+    },
+    [actions.CHART_UPDATE_STOPPED](state) {
+      return Object.assign({}, state,
```
ditto object spread `{ ... }`

williaster commented on superset/assets/javascripts/chart/chartReducer.js:
```diff
@@ -0,0 +1,108 @@
+        });
+    },
+    [actions.CHART_RENDERING_FAILED](state) {
+      return Object.assign({}, state, {
```
ditto object spread `{ ... }`

williaster commented on superset/assets/javascripts/chart/chartReducer.js:
```diff
@@ -0,0 +1,108 @@
+      });
+    },
+    [actions.CHART_UPDATE_TIMEOUT](state) {
+      return Object.assign({}, state, {
```
ditto object spread `{ ... }`

williaster commented on superset/assets/javascripts/chart/chartReducer.js:
```diff
@@ -0,0 +1,108 @@
+      return Object.assign({}, state, {
+        chartStatus: 'failed',
+        chartAlert: (
+        '<strong>Query timeout</strong> - visualization query are set to timeout at ' +
```
typo `query` -> `queries`

williaster commented on superset/assets/javascripts/chart/chartReducer.js:
```diff
@@ -0,0 +1,108 @@
+        '<strong>Query timeout</strong> - visualization query are set to timeout at ' +
+        `${action.timeout} seconds. ` +
+        t('Perhaps your data has grown, your database is under unusual load, ' +
+          'or you are simply querying a data source that is to large ' +
```
typo `to` -> `too`

graceguo-supercat replied:
:_(

williaster commented on superset/assets/javascripts/chart/chartReducer.js:
```diff
@@ -0,0 +1,108 @@
+      });
+    },
+    [actions.CHART_UPDATE_FAILED](state) {
+      return Object.assign({}, state, {
```
ditto object spread `{ ... }`

williaster commented on superset/assets/javascripts/chart/chartReducer.js:
```diff
@@ -0,0 +1,108 @@
+  }
+
+  if (action.type in actionHandlers) {
+    return Object.assign({}, charts, {
```
ditto object spread `{ ... }`

williaster commented on superset/assets/javascripts/components/StackTraceMessage.jsx:
```diff
@@ -0,0 +1,59 @@
+        {this.hasTrace() &&
+        <Collapse in={this.state.showStackTrace}>
+          <pre>
+            {this.props.queryResponse.stacktrace}
```
this would throw without a default `queryResponse` object (or make required)

williaster commented on superset/assets/javascripts/components/StackTraceMessage.jsx:
```diff
@@ -0,0 +1,59 @@
+  render() {
+    const msg = (
+      <div>
+        <p
```
we have used this a couple of places, could create an `HTML` component for it (nbd)

graceguo-supercat replied:
somehow I didn't see this was used in other place...

williaster commented on superset/assets/javascripts/dashboard/actions.js:
```diff
@@ -0,0 +1,112 @@
+        // if server-side reject the overwrite action,
+        // revert to old state
+        dispatch(updateSliceName(slice, oldName));
+        notify.error("You don't have the rights to alter this slice");
```
nit single quote strings, should turn on linting for this

mistercrunch replied:
Linter allows double quotes when string contains a single quote which is good

ljharb replied:
although, typographically, a single quote is incorrect here - it should be a curly quote.

williaster commented on superset/assets/javascripts/dashboard/components/Controls.jsx:
```diff
@@ -36,11 +45,11 @@ class Controls extends React.PureComponent {
+    this.props.onChange();
   }
   render() {
     const dashboard = this.props.dashboard;
```
you could destructure all of the variables used in the calls to `this.props.xxx` below to simplify.

williaster commented on superset/assets/javascripts/dashboard/components/Dashboard.jsx:
```diff
@@ -0,0 +1,334 @@
+
+  // return charts in array
+  getAllSlices() {
+    return Object.keys(this.props.slices).map(key => (this.props.slices[key]));
```
you could also use `Object.values`

williaster commented on superset/assets/javascripts/dashboard/components/Dashboard.jsx:
```diff
@@ -0,0 +1,334 @@
+  }
+
+  getFormDataExtra(slice) {
+    const formDataExtra = Object.assign({}, slice.formData);
```
nit spread `{ ... slice.formData }`

williaster commented on superset/assets/javascripts/dashboard/components/Dashboard.jsx:
```diff
@@ -0,0 +1,334 @@
+
+  fetchSlice(slice, force = false) {
+    return this.props.actions.runQuery(
+      this.getFormDataExtra(slice), force, this.props.timeout, slice.chartKey);
```
nit new line for `)`

williaster commented on superset/assets/javascripts/dashboard/components/Dashboard.jsx:
```diff
@@ -0,0 +1,334 @@
+      for (const col in this.props.filters[key]) {
+        const sliceId = parseInt(key, 10);
+        this.props.actions.addFilter(sliceId, col,
+          this.props.filters[key][col], false, false);
```
nit newline for `)`

williaster commented on superset/assets/javascripts/dashboard/components/Dashboard.jsx:
```diff
@@ -0,0 +1,334 @@
+    let resizeTimer;
+    const dash = this;
+    const allSlices = this.getAllSlices();
+    $(window).on('resize', () => {
```
why use `$` here and not straight `JS`?

williaster commented on superset/assets/javascripts/dashboard/components/Dashboard.jsx:
```diff
@@ -0,0 +1,334 @@
+    const dash = this;
+    const slices = slc || this.getAllSlices();
+    if (!interval) {
+      slices.forEach(slice => (dash.fetchSlice(slice, force)));
```
`forEach`s shouldn't have a return value

williaster commented on superset/assets/javascripts/dashboard/components/Dashboard.jsx:
```diff
@@ -0,0 +1,334 @@
+    }
+    const delay = meta.stagger_refresh ? refreshTime / (slices.length - 1) : 0;
+    slices.forEach((slice, i) => {
+      setTimeout(() => dash.fetchSlice(slice, force), delay * i);
```
nit - should wrap in `{}` so no return value

williaster commented on superset/assets/javascripts/dashboard/components/Dashboard.jsx:
```diff
@@ -0,0 +1,334 @@
+            dashboard={this.props.dashboard}
+            userId={this.props.userId}
+            isStarred={this.props.isStarred}
+            updateDashboardTitle={this.updateDashboardTitle.bind(this)}
```
shouldn't bind in the `render` for any of these `func`s

williaster commented on superset/assets/javascripts/dashboard/components/DashboardContainer.jsx:
```diff
@@ -0,0 +1,29 @@
+}
+
+function mapDispatchToProps(dispatch) {
+  const actions = Object.assign({}, chartActions, dashboardActions);
```
nit object spread `{ ... }`

williaster commented on superset/assets/javascripts/dashboard/components/GridLayout.jsx:
```diff
@@ -14,117 +12,94 @@ const ResponsiveReactGridLayout = WidthProvider(Responsive);
+  timeout: PropTypes.number,
+  onChange: PropTypes.func,
+  getFormDataExtra: PropTypes.func,
+  fetchSlice: PropTypes.func,
```
you should have defaults for all of these non-required funcs / they would throw below

graceguo-supercat replied:
fixed.

williaster commented on superset/assets/javascripts/dashboard/components/GridCell.jsx:
```diff
@@ -0,0 +1,106 @@
+            timeout={this.props.timeout}
+            viz_type={slice.formData.viz_type}
+            addFilter={this.props.addFilter.bind(this, slice.slice_id)}
+            getFilters={() => (this.props.filters[slice.slice_id])}
```
shouldn't define new funcs in the render if possible

williaster commented on superset/assets/javascripts/dashboard/components/GridCell.jsx:
```diff
@@ -0,0 +1,106 @@
+            removeSlice={this.props.removeSlice}
+            updateSliceName={this.props.updateSliceName}
+            toggleExpandSlice={this.props.toggleExpandSlice}
+            forceRefresh={() => this.props.fetchSlice(true)}
```
shouldn't define new funcs in the render if possible

williaster commented on superset/assets/javascripts/dashboard/components/GridCell.jsx:
```diff
@@ -0,0 +1,106 @@
+            width={this.width()}
+            timeout={this.props.timeout}
+            viz_type={slice.formData.viz_type}
+            addFilter={this.props.addFilter.bind(this, slice.slice_id)}
```
shouldn't  bind in the render

williaster commented on superset/assets/javascripts/dashboard/components/GridCell.jsx:
```diff
@@ -0,0 +1,106 @@
+          <input type="hidden" value="false" />
+          <ChartContainer
+            containerId={`slice-container-${slice.slice_id}`}
+            chartKey={this.props.chartKey}
```
you could destructure a lot of vars from `this.props` to reduce the number of `this.props.xxx` calls you make

williaster commented on superset/assets/javascripts/dashboard/components/GridLayout.jsx:
```diff
@@ -134,27 +109,43 @@ class GridLayout extends React.Component {
-              expandedSlices={this.props.dashboard.metadata.expanded_slices}
+              chartKey={'slice_' + slice.slice_id}
+              datasource={this.props.datasources[slice.form_data.datasource]}
+              filters={this.props.filters}
```
you could destructure a lot of vars from `this.props` to reduce the number of `this.props.xxx` calls

williaster commented on superset/assets/javascripts/dashboard/components/GridLayout.jsx:
```diff
@@ -134,27 +109,43 @@ class GridLayout extends React.Component {
+              isLoading={[undefined, 'loading']
+                .indexOf(this.props.charts['slice_' + slice.slice_id].chartStatus) !== -1}
+              toggleExpandSlice={this.props.toggleExpandSlice}
+              fetchSlice={this.props.fetchSlice.bind(this, this.props.charts['slice_' + slice.slice_id])}
```
shouldn't bind in a render

williaster commented on superset/assets/javascripts/dashboard/components/SaveModal.jsx:
```diff
@@ -13,6 +13,9 @@ const propTypes = {
   css: PropTypes.string,
   dashboard: PropTypes.object.isRequired,
   triggerNode: PropTypes.node.isRequired,
+  readFilters: PropTypes.func,
```
you should add defaults or the calls below could throw

graceguo-supercat replied:
fixed.

williaster commented on superset/assets/javascripts/dashboard/components/SliceAdder.jsx:
```diff
@@ -43,7 +45,7 @@ class SliceAdder extends React.Component {
 
   onEnterModal() {
-    const uri = '/sliceaddview/api/read?_flt_0_created_by=' + this.props.dashboard.curUserId;
+    const uri = '/sliceaddview/api/read?_flt_0_created_by=' + this.props.userId;
```
why not use string template literals? I think they're generally preferred over concatenation. should have a lint rule for this

williaster commented on superset/assets/javascripts/dashboard/components/SliceHeader.jsx:
```diff
@@ -0,0 +1,128 @@
+            <EditableTitle
+              title={slice.slice_name}
+              canEdit={!!this.props.updateSliceName}
+              onSaveTitle={this.onSaveTitle.bind(this)}
```
shouldn't bind in the render

williaster commented on superset/assets/javascripts/dashboard/components/SliceHeader.jsx:
```diff
@@ -0,0 +1,128 @@
+              title={slice.slice_name}
+              canEdit={!!this.props.updateSliceName}
+              onSaveTitle={this.onSaveTitle.bind(this)}
+              noPermitTooltip={'You don\'t have the rights to alter this dashboard.'}
```
should this be wrapped in a `t`?

williaster commented on superset/assets/javascripts/dashboard/components/SliceHeader.jsx:
```diff
@@ -0,0 +1,128 @@
+              </a>
+              <a
+                className={`refresh ${isCached ? 'danger' : ''}`}
+                onClick={() => (this.props.forceRefresh())}
```
you should create a `this.handleRefresh` / don't create new funcs in the render (ditto for other clicks below)

williaster commented on superset/assets/javascripts/dashboard/index.jsx:
```diff
@@ -0,0 +1,29 @@
+const store = createStore(
+  rootReducer, initState, compose(applyMiddleware(thunk), initEnhancer(false)));
+
+ReactDOM.render(
```
wohoo redux dashboards! 🎉

williaster commented on superset/assets/javascripts/dashboard/reducers.js:
```diff
@@ -0,0 +1,193 @@
+  delete common.locale;
+  delete common.language_pack;
+
+  const dashboard = Object.assign({}, bootstrapData.dashboard_data);
```
nit object spread `{ ... }`

williaster commented on superset/assets/javascripts/dashboard/reducers.js:
```diff
@@ -0,0 +1,193 @@
+      filters[sliceId] = filterData[key];
+    }
+  } catch (e) {
+    //
```
should we log something here?

williaster commented on superset/assets/javascripts/dashboard/reducers.js:
```diff
@@ -0,0 +1,193 @@
+
+  return {
+    charts: initCharts,
+    dashboard: Object.assign({}, { filters, dashboard, userId: user_id, datasources, common }),
```
nit object spread

williaster commented on superset/assets/javascripts/dashboard/reducers.js:
```diff
@@ -0,0 +1,193 @@
+  const actionHandlers = {
+    [actions.UPDATE_DASHBOARD_TITLE]() {
+      const newDashboard = Object.assign({}, state.dashboard, { dashboard_title: action.title });
+      return Object.assign({}, state, { dashboard: newDashboard });
```
nit object spread for all of these

williaster reviewed (commented):
Yay for redux dashboards! I had a bunch of mostly nit comments but some important ones like `bind`ing in render `func`s, we should add lint rules for some of these things

graceguo-supercat commented on superset/assets/javascripts/chart/Chart.jsx:
```diff
@@ -0,0 +1,190 @@
+    );
+  }
+
+  render_template(s) {
```
this function is called from ERF or data portal users. I want to handle API change later, and probably shorten URL change is also coming soon.

Commit:
more code review fixes
    - linting fix: rename viz_type to vizType, use spread operator replace Object.assign, etc;
    - no binding in callback;
    - resize dashboard should only re-render charts without referch;
    - use Loading component when chart is loading;
    - for each slice in dashboard, when user clicks edit chart or explore chart, open a new window instead of use current window.

Commit:
Merge remote-tracking branch 'upstream/master' into gg-ChartComponent

mistercrunch commented:
LGTM

williaster commented on superset/assets/javascripts/dashboard/components/DashboardContainer.jsx:
```diff
@@ -0,0 +1,29 @@
+  };
+}
+
+export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);
```
at this point this is probably out of the scope of this PR, but we should also try to embrace the redux `container vs component` paradigm [described here](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0).

williaster commented on superset/assets/javascripts/dashboard/components/GridCell.jsx:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  getDescriptionId(slice) {
+    return 'description_' + slice.slice_id;
```
these should be template literals

williaster commented on superset/assets/javascripts/dashboard/components/GridCell.jsx:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  getHeaderId(slice) {
+    return 'header_' + slice.slice_id;
```
these should be template literals

williaster commented on superset/assets/javascripts/dashboard/components/SliceAdder.jsx:
```diff
@@ -65,14 +67,30 @@ class SliceAdder extends React.Component {
+      .fail((error) => {
+        adder.errored = true;
+        adder.setState({
+          errorMsg: t('Sorry, there was an error adding slices to this dashboard: ') +
```
nit this could be a template literal

williaster commented on superset/assets/javascripts/dashboard/components/SliceHeader.jsx:
```diff
@@ -0,0 +1,142 @@
+            />
+          </div>
+          <div className="chart-controls">
+            <div id={'controls_' + slice.slice_id} className="pull-right">
```
nit template literal

williaster commented on superset/assets/javascripts/dashboard/components/SliceHeader.jsx:
```diff
@@ -0,0 +1,142 @@
+                  <i className="fa fa-table" />
+                </TooltipWrapper>
+              </a>
+              <a className="exploreChart" href={this.props.exploreChartUrl} target="_blank">
```
👏 new tab! 🔥

williaster commented on superset/assets/javascripts/dashboard/index.jsx:
```diff
@@ -0,0 +1,29 @@
+const initState = Object.assign({}, getInitialState(bootstrapData));
+
+const store = createStore(
+  rootReducer, initState, compose(applyMiddleware(thunk), initEnhancer(false)));
```
nit would make all of these new lines

williaster commented on superset/assets/javascripts/dashboard/reducers.js:
```diff
@@ -0,0 +1,188 @@
+  const initCharts = {};
+  dashboard.slices.forEach((slice) => {
+    const chartKey = 'slice_' + slice.slice_id;
+    initCharts[chartKey] = { ...chart,
```
nit would add newline before `...chart`

williaster commented on superset/assets/javascripts/dashboard/reducers.js:
```diff
@@ -0,0 +1,188 @@
+import d3 from 'd3';
+
+import charts, { chart } from '../chart/chartReducer';
+import * as actions from './actions';
```
ditto explicit imports are better

williaster commented on superset/assets/javascripts/explore/components/ExploreChartHeader.jsx:
```diff
@@ -0,0 +1,140 @@
+
+const propTypes = {
+  actions: PropTypes.object.isRequired,
+  can_overwrite: PropTypes.bool.isRequired,
```
similar to `viz_type` these ideally would be `camelCase`

williaster reviewed (commented):
Thanks for taking another pass at this! looks great overall, I had a few last small comments
